Notes/Proposals/Explicit-partial-type-inference.md

```diff
@@ -0,0 +1,313 @@
+
+Based on mentioned advantages that `_` placeholder can bring to type inference we feel to add it to C# worthly.
+
+## Scope
+
+Partial type inference can be solved in various ways.
+We chose a feature enabling to hint the compiler by specifying ambitious type arguments and letting the remaining ones on him.
+It aims at cases, where we want just to specify the arity of desired generic method(type) or specify a parameter that is not possible to infer from the context but let the compiler infer the remaining arguments. 
+
+## Design
+
+As we mentioned in the motivation, we see the usage of our proposal in the following places.
+
+1. method call
+2. object construction
+3. variable declaration
+4. casting
+
+### Method call
+
+We would like to somehow mark inferred type arguments. 
+The most natural approach would be to replace the type arguments with some keyword. 
+Let's pick `_` for it for now and show it in the examples.
+
+```csharp
+void foo<T1, T2, T3>(T2 arg) {}
+
+foo<int, _, int>("string"); // _ is marked as "inferred by compiler"
+```
+
+One can suggest not marking it and just leaving it on the compiler. 
+However, it would bring ambiguities between which arguments should be inferred. 
+
+```csharp
+void foo(int arg1, int arg2) {}
+void foo<T1>(T1 arg1, int arg2) {}
+void foo<T1, T2>(T1 arg1, T2 arg2) {}
+
+foo(1,2); // We don't know which overload to choose
+foo<>(1,2); // Still, it is not clear which generic overload should be choose.
+```
+
+Because of mentioned issues, we believe that marking each type argument, what the compiler will infer, is the right way how to do it.
+To make it possible, we would have to change C# grammar
+
+```
+invocation_expression
+    : primary_expression '(' argument_list? ')'
+    : simple_name_u '(' argument_list? ')'
+    : member_access_u '(' argument_list? ')'
+    : null_conditional_member_access_u '(' argument_list? ')'
+    ;
+
+simple_name_u
+    : identifier type_argument_list_u?
+    ;
+
+member_access_u
+    : primary_expression '.' identifier type_argument_list_u?
+    | predefined_type '.' identifier type_argument_list_u?
+    | qualified_alias_member '.' identifier type_argument_list_u?
+    ;
+
+null_conditional_member_access_u
+    : primary_expression '?' '.' identifier type_argument_list?
+      dependent_access* dependent_member_access_u?
+    | primary_expression '?' '.' identifier type_argument_list_u?
+    ;
+    
+dependent_access
+    : '.' identifier type_argument_list?    // member access
+    | '[' argument_list ']'                 // element access
+    | '(' argument_list? ')'                // invocation
+    ;
+
+dependent_member_access_u
+    : '.' identifier type_argument_list_u?    // member access
+    ;
+
+type_argument_list_u
+    : '<' type_arguments_u '>'
+    ;
+
+type_arguments_u
+    : type_argument_u (',' type_argument_u)*
+    ;   
+
+type_argument_u
+    : type | '_'
```

Comment:
Suggest using https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1217-declaration-expressions as a starting point. ctrl+f for "simple discard". This works not by changing the grammar but by saying that if a type is the identifier `_` then it's considered a *simple discard*.

When implementing, it will also be a good idea to look at how discards are represented in patterns, assignments, etc. as a starting point.